File: algorithms/13_swift/sorting/MergeSort.swift
func mergeSort<T: Comparable>(_ arr: [T]) -> [T] {
    guard arr.count > 1 else { return arr }
    let m = arr.count/2
    return merge(mergeSort(Array(arr[..<m])), mergeSort(Array(arr[m...])))
}
func merge<T: Comparable>(_ l: [T], _ r: [T]) -> [T] {
    var res=[T](); var i=0,j=0
    while i<l.count&&j<r.count { if l[i]<=r[j]{res.append(l[i]);i+=1}else{res.append(r[j]);j+=1} }
    return res+l[i...]+r[j...]
}

File: algorithms/07_typescript/sorting/merge_sort.ts
export function mergeSort(arr: number[]): number[] {
  if (arr.length <= 1) return arr;
  const m = arr.length >> 1;
  return merge(mergeSort(arr.slice(0,m)), mergeSort(arr.slice(m)));
}
function merge(l: number[], r: number[]): number[] {
  const res: number[] = []; let [i,j] = [0,0];
  while (i<l.length && j<r.length) res.push(l[i]<=r[j]?l[i++]:r[j++]);
  return [...res,...l.slice(i),...r.slice(j)];
}